backend/Testing/Browser/Page/BasePage.cs

```diff
@@ -46,13 +46,16 @@ public async Task<T> WaitFor()
     {
         if (UrlPattern is not null)
         {
+            //assert to get a good error message
+            await Assertions.Expect(Page).ToHaveURLAsync(UrlPattern);
+            //still wait to make sure we reach the same state we expect
             await Page.WaitForURLAsync(UrlPattern, new() { WaitUntil = WaitUntilState.Load });
         }
         else
         {
             await Page.WaitForLoadStateAsync(LoadState.Load);
         }
-        await Task.WhenAll(TestLocators.Select(l => l.WaitForAsync()));
+        await Task.WhenAll(TestLocators.Select(l => Assertions.Expect(l).ToBeVisibleAsync()));
```

Comment:
this should similarly give a better failure case, though I wasn't able to test it properly. `WaitForAsync` wait's for it to be visible by default.